streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java

```diff
@@ -227,21 +253,72 @@ public final TopologyBuilder addSource(String name, Deserializer keyDeserializer
         if (nodeFactories.containsKey(name))
             throw new TopologyBuilderException("Processor " + name + " is already added.");
 
+        if (topicPattern != null) {
+            throw new TopologyBuilderException("Subscription to topics, partitions and pattern are mutually exclusive");
```

Comment:
We do not need to mention "partition" here since it is supposed to be abstracted from users, ditto below.